src/cellrank/estimators/terminal_states/_gpcca.py

```diff
@@ -532,6 +535,140 @@ def set_initial_states(
         )
         return self
 
+    def get_tsi(self, n_macrostates: int, terminal_states: List[str], cluster_key: str, **kwargs: Any) -> pd.DataFrame:
+        """Compute terminal state identificiation (TSI).
+
+        Parameters
+        ----------
+        n_macrostates
+            Maximum number of macrostates to consider.
+        terminal_states
+            List of terminal states.
+        cluster_key
+            Key in :attr:`~anndata.AnnData.obs` defining cluster labels including terminal states.
+        kwargs
+            Keyword arguments passed to the class' `compute_macrostates` function.
+
+        Returns
+        -------
+        Returns TSI as a Pandas DataFrame and adds the class attribute :attr:`tsi`. The DataFrame contains the columns
+
+        - "Number of macrostates": Number of macrostates computed
+        - "Identified terminal states": Number of terminal states identified
+        - "Optimal identification": Number of terminal states identified when using an optimal identification scheme
+        """
+        macrostates = {}
+        for n_states in range(n_macrostates, 0, -1):
+            self.compute_macrostates(n_states=n_states, cluster_key=cluster_key, **kwargs)
+            macrostates[n_states] = self.macrostates.cat.categories
+
+        max_terminal_states = len(terminal_states)
+
+        tsi_df = collections.defaultdict(list)
+        for n_states, states in macrostates.items():
+            n_terminal_states = (
+                states.str.replace(r"(_).*", "", regex=True).drop_duplicates().isin(terminal_states).sum()
+            )
+            tsi_df["Number of macrostates"].append(n_states)
+            tsi_df["Identified terminal states"].append(n_terminal_states)
+
+            tsi_df["Optimal identification"].append(min(n_states, max_terminal_states))
+
+        tsi_df = pd.DataFrame(tsi_df)
+        self.tsi = tsi_df
+
+        return tsi_df
+
+    def get_tsi_score(self) -> float:
+        """Compute TSI score."""
+        if not hasattr(self, "tsi"):
+            raise RuntimeError("Compute TSI with `get_tsi` first.")
+
+        optimal_score = self.tsi["Optimal identification"].sum()
+
+        return self.tsi["Identified terminal states"].sum() / optimal_score
+
+    @d.dedent
+    def plot_tsi(
+        self,
+        tsi_df: pd.DataFrame,
```

Comment:
Why is `tsi_df` passed here and not being accessed from the class? If it's being saved, this method should access it.
In that case, please check `if self.tsi is None: raise RuntimeError(...)`.